File: sdk-js/src/pith.js
// Pith — context-style discipline for capsule narrative fields.
//
// This is a *normalizer*, not "compression" in the
// information-theoretic sense. It deterministically rewrites a small
// grammar of narrative fields (whitespace normalization, first-N
// sentences, word-boundary truncation with ellipsis) so that cold-
// reading LLMs absorb a consistent, terse context faster.
//
// The discipline is the product. The reference implementation is a
// helper for the discipline. An LLM applying Pith *style* produces
// richer compression than this library can; this library guarantees
// that fields written without LLM judgment still come out terse and
// regular.
//
// Applied automatically by CapsuleBuilder.appendEvent() to event
// payload fields commonly used for narrative:
//   - payload.summary
//   - payload.statement
//   - payload.note
//   - payload.open_items[].item
//   - payload.decisions[].text
//   - payload.milestones[].text
//
// Pass { pith: false } to appendEvent() to opt out, or call
// compressEventPayload() / compressText() directly.

export const PITH_VERSION = "0.6";
const DEFAULT_MAX_CHARS = 280;
const DEFAULT_MAX_SENTENCES = 3;
const ELLIPSIS = "…";

/**
 * compressText(input, options) -> { text, changed, version }
 *
 * options:
 *   maxChars:     number (default 280) — output length cap, including ellipsis
 *   maxSentences: number (default 3)   — sentences kept before length cap
 */
export function compressText(input, options = {}) {
  if (typeof input !== "string") {
    throw new TypeError("compressText: input must be a string");
  }
  const maxChars = positiveIntegerOrDefault(options.maxChars, DEFAULT_MAX_CHARS);
  const maxSentences = positiveIntegerOrDefault(options.maxSentences, DEFAULT_MAX_SENTENCES);

  const normalized = normalizeWhitespace(input);
  const trimmed = firstSentences(normalized, maxSentences);
  const text = truncateAtWordBoundary(trimmed, maxChars);

  return {
    text,
    changed: text !== input,
    version: PITH_VERSION,
  };
}

/**
 * Return a deep-cloned copy of `payload` with known narrative fields
 * normalized. Non-narrative fields (numbers, IDs, hashes, JSON
 * structures) are preserved verbatim.
 */
export function compressEventPayload(payload, options = {}) {
  const copy = cloneJson(payload);
  if (!isRecord(copy)) return copy;
  compressStringField(copy, "summary", options);
  compressStringField(copy, "statement", options);
  compressStringField(copy, "note", options);
  compressTextListField(copy, "open_items", "item", options);
  compressTextListField(copy, "decisions", "text", options);
  compressTextListField(copy, "milestones", "text", options);
  return copy;
}

// ---------- internals ----------

function normalizeWhitespace(input) {
  return input
    .replace(/\r\n?/g, "\n")
    .split("\n")
    .map((line) => line.replace(/[\t ]+/g, " ").trim())
    .filter((line) => line.length > 0)
    .join(" ");
}

function firstSentences(input, maxSentences) {
  if (input.length === 0) return input;
  const matches = input.match(/[^.!?]+(?:[.!?]+|$)/g);
  if (!matches) return input;
  const sentences = matches.map((s) => s.trim()).filter((s) => s.length > 0);
  if (sentences.length <= maxSentences) return input;
  return sentences.slice(0, maxSentences).join(" ");
}

function truncateAtWordBoundary(input, maxChars) {
  if (input.length <= maxChars) return input;
  if (maxChars <= ELLIPSIS.length) return ELLIPSIS.slice(0, maxChars);

  const limit = maxChars - ELLIPSIS.length;
  const prefix = input.slice(0, limit);
  const trimmedPrefix = prefix.trimEnd();
  const lastSpace = trimmedPrefix.lastIndexOf(" ");
  const minimumUsefulBoundary = Math.floor(limit * 0.6);
  const bounded =
    prefix.length !== trimmedPrefix.length
      ? trimmedPrefix
      : lastSpace >= minimumUsefulBoundary
        ? trimmedPrefix.slice(0, lastSpace)
        : trimmedPrefix;
  const cleaned = bounded.replace(/[\s,;:.!?-]+$/u, "");
  return `${cleaned.length > 0 ? cleaned : trimmedPrefix}${ELLIPSIS}`;
}

function positiveIntegerOrDefault(value, fallback) {
  return Number.isInteger(value) && value > 0 ? value : fallback;
}

function compressStringField(record, key, options) {
  if (typeof record[key] !== "string") return;
  record[key] = compressText(record[key], options).text;
}

function compressTextListField(record, listKey, textKey, options) {
  const list = record[listKey];
  if (!Array.isArray(list)) return;
  for (const entry of list) {
    if (isRecord(entry)) compressStringField(entry, textKey, options);
  }
}

function cloneJson(value) {
  if (Array.isArray(value)) return value.map((entry) => cloneJson(entry));
  if (isRecord(value)) {
    const out = {};
    for (const [k, v] of Object.entries(value)) out[k] = cloneJson(v);
    return out;
  }
  return value;
}

function isRecord(value) {
  if (value === null || typeof value !== "object") return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

File: sdk-py/src/capsule/pith.py
"""Pith — context-style discipline for capsule narrative fields.

Direct port of sdk/src/pith.js.
"""

from __future__ import annotations

import copy
import re
from typing import Any

PITH_VERSION = "0.6"
_DEFAULT_MAX_CHARS = 280
_DEFAULT_MAX_SENTENCES = 3
_ELLIPSIS = "…"

_WS_RE = re.compile(r"[\t ]+")
_LINEEND_RE = re.compile(r"\r\n?")
_SENTENCE_RE = re.compile(r"[^.!?]+(?:[.!?]+|$)")
_TRAIL_PUNCT_RE = re.compile(r"[\s,;:.!?\-]+$")


def compress_text(
    input: str, *, max_chars: int | None = None, max_sentences: int | None = None
) -> dict:
    """Compress narrative text by whitespace normalization, sentence limiting, and truncation.

    Returns dict with keys: text, changed, version.
    """
    if not isinstance(input, str):
        raise TypeError("compress_text: input must be a string")
    mc = max_chars if isinstance(max_chars, int) and max_chars > 0 else _DEFAULT_MAX_CHARS
    ms = (
        max_sentences
        if isinstance(max_sentences, int) and max_sentences > 0
        else _DEFAULT_MAX_SENTENCES
    )
    normalized = _normalize_whitespace(input)
    trimmed = _first_sentences(normalized, ms)
    text = _truncate_at_word_boundary(trimmed, mc)
    return {"text": text, "changed": text != input, "version": PITH_VERSION}


def compress_event_payload(payload: Any, **opts: Any) -> Any:
    """Deep-clone payload and normalize known narrative fields.

    Normalizes: summary, statement, note, open_items[].item,
    decisions[].text, milestones[].text.
    Non-dict inputs pass through unchanged.
    """
    copy_ = copy.deepcopy(payload)
    if not isinstance(copy_, dict):
        return copy_
    _compress_field(copy_, "summary", opts)
    _compress_field(copy_, "statement", opts)
    _compress_field(copy_, "note", opts)
    _compress_list_field(copy_, "open_items", "item", opts)
    _compress_list_field(copy_, "decisions", "text", opts)
    _compress_list_field(copy_, "milestones", "text", opts)
    return copy_


def _normalize_whitespace(s: str) -> str:
    """Collapse \\r\\n? → \\n, split lines, collapse [\\t ]+ → space, strip, join."""
    s = _LINEEND_RE.sub("\n", s)
    lines = [_WS_RE.sub(" ", line).strip() for line in s.split("\n")]
    return " ".join(line for line in lines if line)


def _first_sentences(s: str, max_sentences: int) -> str:
    """Return the first N sentences, or original if fewer than N exist."""
    if not s:
        return s
    matches = _SENTENCE_RE.findall(s)
    if not matches:
        return s
    sentences = [m.strip() for m in matches if m.strip()]
    if len(sentences) <= max_sentences:
        return s
    return " ".join(sentences[:max_sentences])


def _truncate_at_word_boundary(s: str, max_chars: int) -> str:
    """Truncate at word boundary with ellipsis, respecting max_chars total length."""
    if len(s) <= max_chars:
        return s
    if max_chars <= len(_ELLIPSIS):
        return _ELLIPSIS[:max_chars]
    limit = max_chars - len(_ELLIPSIS)
    prefix = s[:limit]
    trimmed_prefix = prefix.rstrip()
    last_space = trimmed_prefix.rfind(" ")
    minimum_useful = limit * 6 // 10
    if len(prefix) != len(trimmed_prefix):
        # Prefix had trailing whitespace before rstrip; use rstripped version.
        bounded = trimmed_prefix
    elif last_space >= minimum_useful:
        # Last space is at >= 60% of limit; cut there.
        bounded = trimmed_prefix[:last_space]
    else:
        # Last space is < 60%; use rstripped prefix.
        bounded = trimmed_prefix
    cleaned = _TRAIL_PUNCT_RE.sub("", bounded)
    base = cleaned if cleaned else trimmed_prefix
    return base + _ELLIPSIS


def _compress_field(record: dict, key: str, opts: dict) -> None:
    """Compress a single string field in-place if it exists."""
    if not isinstance(record.get(key), str):
        return
    record[key] = compress_text(record[key], **opts)["text"]


def _compress_list_field(record: dict, list_key: str, text_key: str, opts: dict) -> None:
    """Compress a text field within list entries in-place."""
    lst = record.get(list_key)
    if not isinstance(lst, list):
        return
    for entry in lst:
        if isinstance(entry, dict):
            _compress_field(entry, text_key, opts)
